# 1.0 Introduction


## 1.1 Project Overview

Karamoja, a region in north-eastern Uganda, faces severe food insecurity due to factors like intense droughts, pest infestations, and disease outbreaks. This has resulted in very low grain yields, especially staple crops such as sorghum and maize. In order to overcome such barriers, numerous NGOs offer technical assistance and agricultural inputs to the local farmers but they lack visibility into the overall state of the region and often need to rely on some very local sources of information to prioritize their activities, creating challenges in identifying and targeting preventive efforts.

DDI has been tasked to build a food security monitoring tool that will assist in decision making. The data collected via this tool will enable NGOs to gain a better insight into crop yields and population distribution throughout Karamoja, making resource allocation far more effective.

Through this programme, the Agri-tech team at DDI has devised a technique that can help in remotely measuring crop yield by making use of satellite images. The 2017 crop season was run through the model, and the results are now available for analysis.

## 1.2 Dataset description

The datasets provided for this project include both geospatial data and tabular data:

**Shapefiles:**

-Boundaries of Uganda Subcounties

-Boundaries of Uganda Districts

-Crop Type Map for Sorghum (i.e. position of the Sorghum fields)

-Crop Type Map for Maize

**Tables:**

-Yield and Population per Subcounty

-Yield and Population per District

Each table contains fields such as:

-POP: Total population for the subcounty/district

-S_Yield_Ha: Average yield for sorghum (Kg/Ha)

-M_Yield_Ha: Average yield for maize (Kg/Ha)

-Crop_Area_Ha: Total crop area (Ha)

-S_Area_Ha: Total sorghum crop area (Ha)

-M_Area_Ha: Total maize crop area (Ha)

-S_Prod_Tot: Total sorghum productivity (Kg)

-M_Prod_Tot: Total maize productivity (Kg)

## 1.3 Tools and Libraries

To accomplish the data preparation and cleaning tasks, I will use Google Colab, a cloud-based Python environment. I shall use the following libraries:

**Pandas:** For data manipulation and analysis

**GeoPandas:** For handling geospatial data


After preparing the data, I will use Tableau to create the final interactive visualization tool.

# 2.0 Data Loading and Exploration

## 2.1 Importing the necessary Libraries

In [2]:
# Importing pandas
import pandas as pd

In [3]:
# Install Geopandas
!pip install geopandas

In [4]:
#importing Geopandas
import geopandas as gpd

## 2.2 Loading and exploring the datasets

**Loading the Tabular Data.**

In [5]:
# Loading the Uganda_Karamoja_District_Crop_Yield_Population file
distr_cyp = pd.read_csv("/content/Uganda_Karamoja_District_Crop_Yield_Population.csv")

In [6]:
# Loading the Uganda_Karamoja_Subcounty_Crop_Yield_Population file
subc_cyp = pd.read_csv("/content/Uganda_Karamoja_Subcounty_Crop_Yield_Population.csv")

**Loading the shapefiles**

In [7]:
# Loading the Crop_Type_Map_Maize.shp file
maize_map = gpd.read_file("/content/Crop_Type_Map_Maize.shp")

In [8]:
# Loading the Crop_Type_Map_Sorghum.shp file
sorg_map = gpd.read_file("/content/Crop_Type_Map_Sorghum.shp")

In [9]:
# Loading the Uganda_Districts.shp
ug_dist = gpd.read_file("/content/Uganda_Districts.shp")

In [10]:
# Loading the Uganda_Subcounties.shp
subc_ug = gpd.read_file("/content/Uganda_Subcounties.shp")

## 2.3 Data Inspection

**Inspecting the Tabular data.**

In [11]:
#Identify the size of the dataset the Uganda_Karamoja_District_Crop_Yield_Population
distr_cyp.shape

(7, 11)

In [12]:
#Identify the size of the dataset Uganda_Karamoja_Subcounty_Crop_Yield_Population
subc_cyp.shape

(52, 13)

In [13]:
# Checking to observe the first 10 elements in the
# Uganda_Karamoja_District_Crop_Yield_Population dataset
distr_cyp.head(10)

OBJECTID           NAME     POP        Area  S_Yield_Ha  M_Yield_Ha  \
0        92           ABIM   90385  2771977106         449        1040   
1        96         AMUDAT  101790  1643582836         205        1297   
2        20        KAABONG  627057  7373606003         279         945   
3        85         KOTIDO  243157  3641539808         331        1148   
4         5         MOROTO  127811  3570160948         128         355   
5        54  NAKAPIRIPIRIT  146780  4216323900         356        1264   
6        80          NAPAK  167625  4508782023         137         854   

   Crop_Area_Ha     S_Area_Ha    M_Area_Ha  S_Prod_Tot  M_Prod_Tot  
0   5470.068394   3277.295971  1848.621855     1471506     1922567  
1   5765.443719   2973.423860  2733.661014      609552     3545558  
2  28121.672530  20544.194960  7394.416334     5731830     6987723  
3  53032.649450  50247.443900  1751.372284    16631904     2010575  
4   5954.814048   4741.748776  1190.050606      606944      422468  
5  26372.698490  19237.333210  6425.788414     6848491     8122197  
6  22944.296020  16142.015880  6543.719066     2211456     5588336

In [14]:
# Checking to observe the first 10 elements in the
# Uganda_Karamoja_Subcounty_Crop_Yield_Population dataset
subc_cyp.head(10)

OBJECTID        SUBCOUNTY_NAME DISTRICT_NAME    POP        Area Karamoja  \
0       263               KACHERI        KOTIDO  17244  1067176155        Y   
1       264                KOTIDO        KOTIDO  52771   597575188        Y   
2       265   KOTIDO TOWN COUNCIL        KOTIDO  27389    23972401        Y   
3       266          NAKAPERIMORU        KOTIDO  38775   419111591        Y   
4       267            PANYANGARA        KOTIDO  65704   880955930        Y   
5       268                RENGEN        KOTIDO  41273   652744859        Y   
6       591         KAABONG  EAST       KAABONG  42221    60801942        Y   
7       592  KAABONG TOWN COUNCIL       KAABONG  38857    13071455        Y   
8       593          KAABONG WEST       KAABONG  41454    67612362        Y   
9       594              KALAPATA       KAABONG  99203   223116860        Y   

   S_Yield_Ha   M_Yield_Ha  Crop_Area_Ha     S_Area_Ha   M_Area_Ha  \
0  354.207411  1137.467019   7023.533691   6434.342449  528.124229   
1  367.890523  1162.996687  13587.990760  12455.592640  824.767081   
2  369.314177  1167.005832   1656.531855   1520.322052    8.561644   
3  283.324569   852.366578   7087.823334   6761.488901   45.721712   
4  373.836926  1283.859882  10398.249390  10111.198130  172.611914   
5  271.185843  1047.180128  13278.520430  12964.499730  171.585704   
6  160.588525   650.123565   1021.316990   1004.647534   15.660409   
7  212.210703  1229.213444    175.080527    156.503888   18.526186   
8  208.040518  1056.416068    967.336727    910.254122   56.275355   
9  209.809352   633.423182   2834.349905   2501.368282  331.225795   

     S_Prod_Tot     M_Prod_Tot  
0  2.279092e+06  600723.892900  
1  4.582294e+06  959201.382500  
2  5.614765e+05    9991.488268  
3  1.915696e+06   38971.659080  
4  3.779939e+06  221609.511400  
5  3.515789e+06  179681.139000  
6  1.613349e+05   10181.201000  
7  3.321180e+04   22772.636430  
8  1.893697e+05   59450.189420  
9  5.248105e+05  209806.097000

In [15]:
# Any additional info.
distr_cyp.info

<bound method DataFrame.info of    OBJECTID           NAME     POP        Area  S_Yield_Ha  M_Yield_Ha  \
0        92           ABIM   90385  2771977106         449        1040   
1        96         AMUDAT  101790  1643582836         205        1297   
2        20        KAABONG  627057  7373606003         279         945   
3        85         KOTIDO  243157  3641539808         331        1148   
4         5         MOROTO  127811  3570160948         128         355   
5        54  NAKAPIRIPIRIT  146780  4216323900         356        1264   
6        80          NAPAK  167625  4508782023         137         854   

   Crop_Area_Ha     S_Area_Ha    M_Area_Ha  S_Prod_Tot  M_Prod_Tot  
0   5470.068394   3277.295971  1848.621855     1471506     1922567  
1   5765.443719   2973.423860  2733.661014      609552     3545558  
2  28121.672530  20544.194960  7394.416334     5731830     6987723  
3  53032.649450  50247.443900  1751.372284    16631904     2010575  
4   5954.814048   4741.748776  1190.050606      606944      422468  
5  26372.698490  19237.333210  6425.788414     6848491     8122197  
6  22944.296020  16142.015880  6543.719066     2211456     5588336  >

In [16]:
# Any additional info.
subc_cyp.info

<bound method DataFrame.info of     OBJECTID              SUBCOUNTY_NAME  DISTRICT_NAME     POP        Area  \
0        263                     KACHERI         KOTIDO   17244  1067176155   
1        264                      KOTIDO         KOTIDO   52771   597575188   
2        265         KOTIDO TOWN COUNCIL         KOTIDO   27389    23972401   
3        266                NAKAPERIMORU         KOTIDO   38775   419111591   
4        267                  PANYANGARA         KOTIDO   65704   880955930   
5        268                      RENGEN         KOTIDO   41273   652744859   
6        591               KAABONG  EAST        KAABONG   42221    60801942   
7        592        KAABONG TOWN COUNCIL        KAABONG   38857    13071455   
8        593                KAABONG WEST        KAABONG   41454    67612362   
9        594                    KALAPATA        KAABONG   99203   223116860   
10       595                      KAMION        KAABONG   60070  1199409465   
11       596                      KAPEDO        KAABONG   33951   304557831   
12       597                     KATHILE        KAABONG   56870   228457417   
13       598                   KAWALAKOL        KAABONG   37691  1742041261   
14       599                  LOBALANGIT        KAABONG   22907   302572789   
15       600                      LODIKO        KAABONG   31820   397700845   
16       601                     LOLELIA        KAABONG  100919  1069331387   
17       602                      LOYORO        KAABONG   16608  1330019236   
18       603            NAPORE (KARENGA)        KAABONG   20394   309577387   
19       604                       SIDOK        KAABONG   24092   125330583   
20       765                        ABIM           ABIM   14377   188613057   
21       766           ABIM TOWN COUNCIL           ABIM   11489   115941152   
22       767                      ALEREK           ABIM   12573  1255760525   
23       768                     LOTUKEI           ABIM   18641   191684465   
24       769                     MORULEM           ABIM   16834   250364895   
25       810                      KARITA         AMUDAT   45015   658008801   
26       811                       LOROO         AMUDAT   22541   371970302   
27       812                      AMUDAT         AMUDAT   30181   596393139   
28       813         AMUDAT TOWN COUNCIL         AMUDAT    4053    17207110   
29       925                  KATIKEKILE         MOROTO   11772   277451997   
30       926                    NADUNGET         MOROTO   42115   596023780   
31       927                       TAPAC         MOROTO   21999   622266516   
32       928           NORTHERN DIVISION         MOROTO    7161     2738777   
33       929           SOUTHERN DIVISION         MOROTO    3272     2121209   
34       978                     LOREGAE  NAKAPIRIPIRIT   26644   928683431   
35       979  NAKAPIRIPIRIT TOWN COUNCIL  NAKAPIRIPIRIT    1418     4673094   
36       980                      NAMALU  NAKAPIRIPIRIT   25905   553067364   
37       981                 KAKOMONGOLE  NAKAPIRIPIRIT   15960   312709097   
38       982                    LOLACHAT  NAKAPIRIPIRIT   22801   642401705   
39       983                 LORENGEDWAT  NAKAPIRIPIRIT    9929   435714604   
40       984                   NABILATUK  NAKAPIRIPIRIT   27534   499776866   
41      1023                      IRIIRI          NAPAK   46979  1030623258   
42      1024                      LOPEEI          NAPAK   18815   790782350   
43      1025                 LORENGECORA          NAPAK    2503   118997955   
44      1026                      LOTOME          NAPAK   17834   125857565   
45      1027                      MATANY          NAPAK   29495   338252511   
46      1028          NAPAK TOWN COUNCIL          NAPAK    8010   150906375   
47      1029                   NGOLERIET          NAPAK   20790   158887552   
48      1295                     NYAKWAE           ABIM   16470   769609687   
49      1313      

**Inspecting the shapefiles**

In [17]:
#Identify the size of the datasets in rows and columns
print(maize_map.shape)
print(sorg_map.shape)
print(ug_dist.shape)
print(subc_ug.shape)

(28233, 2)
(36424, 2)
(112, 3)
(1382, 3)


In [18]:
#Cbserve the first 10 elements in the dataset
maize_map.head(10)

DN                                           geometry
0   6  POLYGON ((609732.781 467760.000, 609720.000 46...
1   6  POLYGON ((609840.000 467661.872, 609840.000 46...
2   6  POLYGON ((606540.000 464340.000, 606600.000 46...
3   6  POLYGON ((614040.000 461640.000, 614160.000 46...
4   6  POLYGON ((613320.000 460680.000, 613380.000 46...
5   6  POLYGON ((607740.000 460380.000, 607800.000 46...
6   6  POLYGON ((609480.000 460260.000, 609540.000 46...
7   6  POLYGON ((609180.000 460140.000, 609300.000 46...
8   6  POLYGON ((607860.000 460020.000, 607920.000 46...
9   6  POLYGON ((610860.000 459840.000, 610920.000 45...

In [19]:
#Cbserve the first 10 elements in the dataset
sorg_map.head(10)

DN                                           geometry
0  212  POLYGON ((609120.000 466020.000, 609180.000 46...
1  212  POLYGON ((608820.000 462780.000, 608880.000 46...
2  212  POLYGON ((607620.000 460020.000, 607680.000 46...
3  212  POLYGON ((606360.000 458280.000, 606420.000 45...
4  212  POLYGON ((601260.000 457560.000, 601320.000 45...
5  212  POLYGON ((601140.000 457500.000, 601200.000 45...
6  212  POLYGON ((601140.000 457380.000, 601320.000 45...
7  212  POLYGON ((601920.000 457260.000, 601980.000 45...
8  212  POLYGON ((602760.000 456840.000, 602820.000 45...
9  212  POLYGON ((601860.000 456660.000, 601920.000 45...

In [20]:
#Cbserve the first 10 elements in the dataset
ug_dist.head(10)

OBJECTID DNAME2014                                           geometry
0         1    MASAKA  POLYGON ((32.00051 -0.60761, 32.00053 -0.60993...
1         2  ALEBTONG  POLYGON ((33.41593 2.14287, 33.41520 2.14219, ...
2         3   BUKEDEA  POLYGON ((34.20871 1.23683, 34.20847 1.23637, ...
3         4     BUSIA  POLYGON ((34.01262 0.24410, 34.01199 0.24363, ...
4         5    MOROTO  POLYGON ((34.90168 2.56680, 34.90168 2.56680, ...
5         6  KABAROLE  POLYGON ((30.37034 0.49085, 30.37034 0.49085, ...
6         7     ZOMBO  POLYGON ((31.00337 2.55750, 31.00337 2.55750, ...
7         8    BUDAKA  POLYGON ((33.99608 0.97423, 33.99608 0.97423, ...
8         9   MARACHA  POLYGON ((30.97223 3.17795, 30.97195 3.17673, ...
9        10    TORORO  POLYGON ((34.28050 0.66552, 34.28098 0.66587, ...

In [21]:
#Cbserve the first 10 elements in the dataset
subc_ug.head(10)

OBJECTID                 SNAME2014  \
0      1001                   KAKOOGE   
1      1002                   KALUNGI   
2      1003                 LWABIYATA   
3      1004                 LWAMPANGA   
4      1005                NABISWEERA   
5      1006                  NAKITOMA   
6      1007                WABINYONYI   
7      1008  NAKASONGOLA TOWN COUNCIL   
8      1009                     BANDA   
9      1010                   BUHEMBA   

                                            geometry  
0  POLYGON ((32.36852 1.28129, 32.36862 1.28116, ...  
1  POLYGON ((32.65120 1.43626, 32.65483 1.43618, ...  
2  POLYGON ((32.44624 1.46113, 32.44397 1.45635, ...  
3  POLYGON ((32.51158 1.52129, 32.51195 1.52123, ...  
4  POLYGON ((32.30877 1.25890, 32.30877 1.25890, ...  
5  POLYGON ((32.16356 1.64770, 32.17149 1.63502, ...  
6  POLYGON ((32.56476 1.10261, 32.55736 1.10493, ...  
7  POLYGON ((32.45697 1.35290, 32.45739 1.35273, ...  
8  POLYGON ((33.90403 0.27950, 33.90438 0.27725, ...  
9  MULTIPOLYGON (((33.71809 0.30086, 33.71874 0.3...

In [22]:
# Any additional info.
maize_map.info

<bound method DataFrame.info of        DN                                           geometry
0       6  POLYGON ((609732.781 467760.000, 609720.000 46...
1       6  POLYGON ((609840.000 467661.872, 609840.000 46...
2       6  POLYGON ((606540.000 464340.000, 606600.000 46...
3       6  POLYGON ((614040.000 461640.000, 614160.000 46...
4       6  POLYGON ((613320.000 460680.000, 613380.000 46...
...    ..                                                ...
28228   6  POLYGON ((698520.000 151560.000, 698580.000 15...
28229   6  POLYGON ((699120.000 151620.000, 699180.000 15...
28230   6  POLYGON ((699000.000 151500.000, 699060.000 15...
28231   6  POLYGON ((699000.000 152160.000, 699060.000 15...
28232   6  POLYGON ((698820.000 151200.000, 698940.000 15...

[28233 rows x 2 columns]>

In [23]:
# Any additional info.
sorg_map.info

<bound method DataFrame.info of         DN                                           geometry
0      212  POLYGON ((609120.000 466020.000, 609180.000 46...
1      212  POLYGON ((608820.000 462780.000, 608880.000 46...
2      212  POLYGON ((607620.000 460020.000, 607680.000 46...
3      212  POLYGON ((606360.000 458280.000, 606420.000 45...
4      212  POLYGON ((601260.000 457560.000, 601320.000 45...
...    ...                                                ...
36419  212  POLYGON ((698220.000 153360.000, 698280.000 15...
36420  212  POLYGON ((698160.000 153300.000, 698220.000 15...
36421  212  POLYGON ((697920.000 152580.000, 697980.000 15...
36422  212  POLYGON ((700620.000 152520.000, 700740.000 15...
36423  212  POLYGON ((699060.000 151620.000, 699120.000 15...

[36424 rows x 2 columns]>

In [24]:
# Any additional info.
ug_dist.info

<bound method DataFrame.info of      OBJECTID  DNAME2014                                           geometry
0           1     MASAKA  POLYGON ((32.00051 -0.60761, 32.00053 -0.60993...
1           2   ALEBTONG  POLYGON ((33.41593 2.14287, 33.41520 2.14219, ...
2           3    BUKEDEA  POLYGON ((34.20871 1.23683, 34.20847 1.23637, ...
3           4      BUSIA  POLYGON ((34.01262 0.24410, 34.01199 0.24363, ...
4           5     MOROTO  POLYGON ((34.90168 2.56680, 34.90168 2.56680, ...
..        ...        ...                                                ...
107       108  BULAMBULI  POLYGON ((34.53499 1.17410, 34.53499 1.17410, ...
108       109      GOMBA  POLYGON ((31.88248 0.09446, 31.88248 0.09446, ...
109       110      RAKAI  POLYGON ((32.00050 -0.82008, 32.00050 -0.82008...
110       111    KANUNGU  POLYGON ((29.85243 -0.93522, 29.85224 -0.93572...
111       112     KIBUKU  POLYGON ((33.82648 0.94900, 33.82251 0.94959, ...

[112 rows x 3 columns]>

In [25]:
# Any additional info.
subc_ug.info

<bound method DataFrame.info of       OBJECTID             SNAME2014  \
0         1001               KAKOOGE   
1         1002               KALUNGI   
2         1003             LWABIYATA   
3         1004             LWAMPANGA   
4         1005            NABISWEERA   
...        ...                   ...   
1377       996    NGOMA TOWN COUNCIL   
1378       997                SEMUTO   
1379       998   SEMUTO TOWN COUNCIL   
1380       999               WAKYATO   
1381      1000  MIGEERA TOWN COUNCIL   

                                               geometry  
0     POLYGON ((32.36852 1.28129, 32.36862 1.28116, ...  
1     POLYGON ((32.65120 1.43626, 32.65483 1.43618, ...  
2     POLYGON ((32.44624 1.46113, 32.44397 1.45635, ...  
3     POLYGON ((32.51158 1.52129, 32.51195 1.52123, ...  
4     POLYGON ((32.30877 1.25890, 32.30877 1.25890, ...  
...                                                 ...  
1377  POLYGON ((32.01615 1.19187, 32.01627 1.19163, ...  
1378  MULTIPOLYGON (((32.22753 0.65549, 32.22840 0.6...  
1379  POLYGON ((32.32632 0.66921, 32.32759 0.66288, ...  
1380  MULTIPOLYGON (((32.30735 0.87261, 32.30766 0.8...  
1381  POLYGON ((32.30295 1.43740, 32.30141 1.43251, ...  

[1382 rows x 3 columns]>

## 2.4 Summary of initial findings.

Preliminarily investigating the datasets reveals that there is tabular data with numerical and categorical variables of population, crop yields, area by district & sub-county. The shapefiles consist of spatial data with geometries showing the boundaries of districts, sub-counties, and crop fields for maize and sorghum. Datasets are structured and ready for further analysis.

# 3.0 Data cleaning

Now let us proceed to clean our data and ensure it can be ready to be worked on in Tableau.


## 3.1 Assessing data completeness

In [26]:
#Checking any missing values for the tabular datasets
print(distr_cyp.isnull().any())

print(subc_cyp.isnull().any())

OBJECTID        False
NAME            False
POP             False
Area            False
S_Yield_Ha      False
M_Yield_Ha      False
Crop_Area_Ha    False
S_Area_Ha       False
M_Area_Ha       False
S_Prod_Tot      False
M_Prod_Tot      False
dtype: bool
OBJECTID          False
SUBCOUNTY_NAME    False
DISTRICT_NAME     False
POP               False
Area              False
Karamoja          False
S_Yield_Ha        False
M_Yield_Ha        False
Crop_Area_Ha      False
S_Area_Ha         False
M_Area_Ha         False
S_Prod_Tot        False
M_Prod_Tot        False
dtype: bool


In [27]:
#Checking any missing values for the shapefiles
print(maize_map.isnull().any())
print(sorg_map.isnull().any())

DN          False
geometry    False
dtype: bool
DN          False
geometry    False
dtype: bool


In [28]:
print(ug_dist.isnull().any())
print(subc_ug.isnull().any())

OBJECTID     False
DNAME2014    False
geometry     False
dtype: bool
OBJECTID     False
SNAME2014    False
geometry     False
dtype: bool


## 3.2 Check for any duplicates

In [29]:
print(distr_cyp.duplicated())
print(subc_cyp.duplicated())
print(maize_map.duplicated())
print(sorg_map.duplicated())
print(ug_dist.duplicated())
print(subc_ug.duplicated())

0    False
1    False
2    False
3    False
4    False
5    False
6    False
dtype: bool
0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35    False
36    False
37    False
38    False
39    False
40    False
41    False
42    False
43    False
44    False
45    False
46    False
47    False
48    False
49    False
50    False
51    False
dtype: bool
0        False
1        False
2        False
3        False
4        False
         ...  
28228    False
28229    False
28230    False
28231    False
28232    False
Length: 28233, dtype: bool
0        False
1        False
2        False
3        False
4        False
        

## 3.3 Ensuring data uniformity

**Changing column names to lowercase**

In [30]:
# Changing the column names to lower case in the tabular datasets
distr_cyp.columns = distr_cyp.columns.str.lower()
subc_cyp.columns = subc_cyp.columns.str.lower()

In [31]:
# Changing the column names to lower case in the shapefiles (GeoDataFrames)
ug_dist.columns = ug_dist.columns.str.lower()
subc_ug.columns = subc_ug.columns.str.lower()
maize_map.columns = maize_map.columns.str.lower()
sorg_map.columns = sorg_map.columns.str.lower()

In [32]:
# Check the results
print(distr_cyp.columns)
print(subc_cyp.columns)

Index(['objectid', 'name', 'pop', 'area', 's_yield_ha', 'm_yield_ha',
       'crop_area_ha', 's_area_ha', 'm_area_ha', 's_prod_tot', 'm_prod_tot'],
      dtype='object')
Index(['objectid', 'subcounty_name', 'district_name', 'pop', 'area',
       'karamoja', 's_yield_ha', 'm_yield_ha', 'crop_area_ha', 's_area_ha',
       'm_area_ha', 's_prod_tot', 'm_prod_tot'],
      dtype='object')


**Checking the Uniformity in Column Values**

In [33]:
unq_dstr = distr_cyp['name']
print(unq_dstr)

0             ABIM
1           AMUDAT
2          KAABONG
3           KOTIDO
4           MOROTO
5    NAKAPIRIPIRIT
6            NAPAK
Name: name, dtype: object


In [34]:
unq_dstr_2 = subc_cyp['district_name'].unique()
print(unq_dstr_2)

['KOTIDO' 'KAABONG' 'ABIM' 'AMUDAT' 'MOROTO' 'NAKAPIRIPIRIT' 'NAPAK']


In [35]:
unq_s = subc_cyp['subcounty_name'].unique()
print(unq_s)

['KACHERI' 'KOTIDO' 'KOTIDO TOWN COUNCIL' 'NAKAPERIMORU' 'PANYANGARA'
 'RENGEN' 'KAABONG  EAST' 'KAABONG TOWN COUNCIL' 'KAABONG WEST' 'KALAPATA'
 'KAMION' 'KAPEDO' 'KATHILE' 'KAWALAKOL' 'LOBALANGIT' 'LODIKO' 'LOLELIA'
 'LOYORO' 'NAPORE (KARENGA)' 'SIDOK' 'ABIM' 'ABIM TOWN COUNCIL' 'ALEREK'
 'LOTUKEI' 'MORULEM' 'KARITA' 'LOROO' 'AMUDAT' 'AMUDAT TOWN COUNCIL'
 'KATIKEKILE' 'NADUNGET' 'TAPAC' 'NORTHERN DIVISION' 'SOUTHERN DIVISION'
 'LOREGAE' 'NAKAPIRIPIRIT TOWN COUNCIL' 'NAMALU' 'KAKOMONGOLE' 'LOLACHAT'
 'LORENGEDWAT' 'NABILATUK' 'IRIIRI' 'LOPEEI' 'LORENGECORA' 'LOTOME'
 'MATANY' 'NAPAK TOWN COUNCIL' 'NGOLERIET' 'NYAKWAE' 'LOKOPO' 'RUPA'
 'MORUITA']


In [36]:
dnm = ug_dist['dname2014'].unique()
print(dnm)

['MASAKA' 'ALEBTONG' 'BUKEDEA' 'BUSIA' 'MOROTO' 'KABAROLE' 'ZOMBO'
 'BUDAKA' 'MARACHA' 'TORORO' 'BUHWEJU' 'KALANGALA' 'BUIKWE' 'LUWERO'
 'KOLE' 'AMOLATAR' 'LAMWO' 'IBANDA' 'WAKISO' 'KAABONG' 'NAKASONGOLA'
 'KABERAMAIDO' 'NAMAYINGO' 'RUKUNGIRI' 'MUBENDE' 'BUVUMA' 'KISORO' 'LUUKA'
 'SIRONKO' 'KALIRO' 'PADER' 'KWEEN' 'AMURU' 'KITGUM' 'KYEGEGWA' 'SERERE'
 'MOYO' 'MASINDI' 'KYENJOJO' 'ISINGIRO' 'JINJA' 'KAPCHORWA' 'BUKWO'
 'AMURIA' 'SHEEMA' 'YUMBE' 'KASESE' 'RUBIRIZI' 'ADJUMANI' 'BUYENDE'
 'KAMULI' 'NTUNGAMO' 'MBARARA' 'NAKAPIRIPIRIT' 'GULU' 'BUKOMANSIMBI'
 'KIBAALE' 'HOIMA' 'MPIGI' 'DOKOLO' 'MITOOMA' 'LWENGO' 'KAYUNGA'
 'LYANTONDE' 'SSEMBABULE' 'BUDUDA' 'AGAGO' 'APAC' 'PALLISA' 'LIRA'
 'NAKASEKE' 'KUMI' 'BUTAMBALA' 'KALUNGU' 'KATAKWI' 'BULIISA' 'KAMWENGE'
 'MANAFWA' 'KYANKWANZI' 'NAPAK' 'KIRUHURA' 'NAMUTUMBA' 'KIBOGA' 'BUGIRI'
 'KOTIDO' 'ARUA' 'SOROTI' 'NEBBI' 'KAMPALA' 'MBALE' 'OTUKE' 'ABIM' 'NGORA'
 'MUKONO' 'KABALE' 'AMUDAT' 'MAYUGE' 'BUSHENYI' 'IGANGA' 'MITYANA'
 'BUTALEJA' 'NTOROKO' '

In [37]:
snm = subc_ug['sname2014'].unique()
print(snm)

['KAKOOGE' 'KALUNGI' 'LWABIYATA' ... 'SEMUTO TOWN COUNCIL' 'WAKYATO'
 'MIGEERA TOWN COUNCIL']


**Renaming Columns**

For this I plan on renaming some columns in the datasets to ensure uniformity among similar data and to ensure an easier time when merging the data.

In [38]:
#I want to ensure all distict categories are under the variable distict_name
distr_cyp.rename(columns={'name': 'district_name'}, inplace=True)

In [39]:
ug_dist.rename(columns={'dname2014': 'district_name'}, inplace=True)

In [40]:
#The same for the sub-counties column to be under the variable subcounty_name
subc_ug.rename(columns={'sname2014': 'subcounty_name'}, inplace=True)

In [41]:
#Review the datasets
print(distr_cyp.columns)
print(subc_cyp.columns)
print(ug_dist.columns)
print(subc_ug.columns)

Index(['objectid', 'district_name', 'pop', 'area', 's_yield_ha', 'm_yield_ha',
       'crop_area_ha', 's_area_ha', 'm_area_ha', 's_prod_tot', 'm_prod_tot'],
      dtype='object')
Index(['objectid', 'subcounty_name', 'district_name', 'pop', 'area',
       'karamoja', 's_yield_ha', 'm_yield_ha', 'crop_area_ha', 's_area_ha',
       'm_area_ha', 's_prod_tot', 'm_prod_tot'],
      dtype='object')
Index(['objectid', 'district_name', 'geometry'], dtype='object')
Index(['objectid', 'subcounty_name', 'geometry'], dtype='object')


# 4.0 Data Integration and Merging

## 4.1. Merging Tabular Data with Shapefiles

In [42]:
# Merging district-level crop yield data with district shapefile
district_merged = pd.merge(ug_dist, distr_cyp, on='district_name', how='inner')

In [43]:
# Drop 'objectid_y' and rename 'objectid_x' to 'objectid'
district_merged = district_merged.drop(columns=['objectid_y']).rename(columns={'objectid_x': 'objectid'})

In [44]:
district_merged.head(2)

objectid district_name                                           geometry  \
0         5        MOROTO  POLYGON ((34.90168 2.56680, 34.90168 2.56680, ...   
1        20       KAABONG  POLYGON ((34.39005 3.48760, 34.39005 3.48760, ...   

      pop        area  s_yield_ha  m_yield_ha  crop_area_ha     s_area_ha  \
0  127811  3570160948         128         355   5954.814048   4741.748776   
1  627057  7373606003         279         945  28121.672530  20544.194960   

     m_area_ha  s_prod_tot  m_prod_tot  
0  1190.050606      606944      422468  
1  7394.416334     5731830     6987723

In [45]:
# Merging sub-county-level crop yield data with sub-county shapefile
subcounty_merged = pd.merge(subc_ug, subc_cyp, on='subcounty_name', how='inner')

In [46]:
# Drop 'objectid_y' and rename 'objectid_x' to 'objectid'
subcounty_merged = subcounty_merged.drop(columns=['objectid_y']).rename(columns={'objectid_x': 'objectid'})

In [47]:
subcounty_merged.head(3)

objectid subcounty_name                                           geometry  \
0      1023         IRIIRI  POLYGON ((34.51034 2.31072, 34.52164 2.30961, ...   
1      1024         LOPEEI  POLYGON ((34.28353 2.90633, 34.28374 2.90631, ...   
2      1025    LORENGECORA  MULTIPOLYGON (((34.22260 2.19440, 34.22280 2.1...   

  district_name    pop        area karamoja  s_yield_ha   m_yield_ha  \
0         NAPAK  46979  1030623258        Y  229.299157   873.452242   
1         NAPAK  18815   790782350        Y  108.156411  1053.257168   
2         NAPAK   2503   118997955        Y  184.081197   841.815010   

   crop_area_ha    s_area_ha    m_area_ha   s_prod_tot    m_prod_tot  
0   5473.443674  1550.944570  3840.698081  355630.2829  3.354666e+06  
1   3120.318045  2121.747320   955.264617  229480.5759  1.006139e+06  
2   1031.196285   672.121886   358.550335  123725.0013  3.018331e+05

## 4.2 Merging all the CSV tabular datasets

In [48]:
merged_cyp = pd.merge(distr_cyp, subc_cyp, on='district_name', how='inner')

In [57]:
merged_cyp.columns

Index(['objectid_x', 'district_name', 'pop_x', 'area_x', 's_yield_ha_x',
       'm_yield_ha_x', 'crop_area_ha_x', 's_area_ha_x', 'm_area_ha_x',
       's_prod_tot_x', 'm_prod_tot_x', 'objectid_y', 'subcounty_name', 'pop_y',
       'area_y', 'karamoja', 's_yield_ha_y', 'm_yield_ha_y', 'crop_area_ha_y',
       's_area_ha_y', 'm_area_ha_y', 's_prod_tot_y', 'm_prod_tot_y'],
      dtype='object')

In [58]:
#Remove the objectid_y and rename objectid_x to objectid
merged_cyp = merged_cyp.drop(columns=['objectid_y'])
merged_cyp = merged_cyp.rename(columns={'objectid_x': 'objectid'})

In [59]:
merged_cyp.head(5)

objectid district_name  pop_x      area_x  s_yield_ha_x  m_yield_ha_x  \
0        92          ABIM  90385  2771977106           449          1040   
1        92          ABIM  90385  2771977106           449          1040   
2        92          ABIM  90385  2771977106           449          1040   
3        92          ABIM  90385  2771977106           449          1040   
4        92          ABIM  90385  2771977106           449          1040   

   crop_area_ha_x  s_area_ha_x  m_area_ha_x  s_prod_tot_x  ...  pop_y  \
0     5470.068394  3277.295971  1848.621855       1471506  ...  14377   
1     5470.068394  3277.295971  1848.621855       1471506  ...  11489   
2     5470.068394  3277.295971  1848.621855       1471506  ...  12573   
3     5470.068394  3277.295971  1848.621855       1471506  ...  18641   
4     5470.068394  3277.295971  1848.621855       1471506  ...  16834   

       area_y  karamoja  s_yield_ha_y m_yield_ha_y  crop_area_ha_y  \
0   188613057         Y    501.263249  1183.089648      999.371589   
1   115941152         Y    439.750995  1061.151087      788.462800   
2  1255760525         Y    520.407672  1259.122045     1266.109868   
3   191684465         Y    382.868986  1003.303910      698.171218   
4   250364895         Y    407.500416   962.958170      925.054130   

   s_area_ha_y  m_area_ha_y  s_prod_tot_y  m_prod_tot_y  
0   593.531315   259.449330   297515.4356   306951.8167  
1   472.601700   282.613094   207827.0681   299895.1917  
2   902.684120   354.449035   469763.7415   446294.5944  
3   364.988891   321.847331   139742.9265   322910.6857  
4   600.673107   299.151724   244774.5413   288070.5968  

[5 rows x 22 columns]

# 5.0 Data Analysis

## 5.1 Summary Statistics

In [49]:
district_summary = district_merged.describe()
print(district_summary)

        objectid            pop          area  s_yield_ha   m_yield_ha  \
count   7.000000       7.000000  7.000000e+00    7.000000     7.000000   
mean   61.714286  214943.571429  3.960853e+09  269.285714   986.142857   
std    36.481567  188604.280916  1.781860e+09  119.243049   321.566700   
min     5.000000   90385.000000  1.643583e+09  128.000000   355.000000   
25%    37.000000  114800.500000  3.171069e+09  171.000000   899.500000   
50%    80.000000  146780.000000  3.641540e+09  279.000000  1040.000000   
75%    88.500000  205391.000000  4.362553e+09  343.500000  1206.000000   
max    96.000000  627057.000000  7.373606e+09  449.000000  1297.000000   

       crop_area_ha     s_area_ha    m_area_ha    s_prod_tot    m_prod_tot  
count      7.000000      7.000000     7.000000  7.000000e+00  7.000000e+00  
mean   21094.520379  16737.636651  3983.947082  4.873098e+06  4.085632e+06  
std    17363.854165  16625.963460  2678.911441  5.743724e+06  2.877188e+06  
min     5470.068394   297

In [50]:
subcounty_summary = subcounty_merged.describe()
print(subcounty_summary)

          objectid            pop          area  s_yield_ha   m_yield_ha  \
count    59.000000      59.000000  5.900000e+01   59.000000    59.000000   
mean    791.135593   26153.627119  4.702248e+08  256.422010   845.078821   
std     277.920198   21016.738870  4.921699e+08  121.448919   401.435681   
min     263.000000    1418.000000  2.121209e+06  108.156411     0.000000   
25%     598.500000   11630.500000  9.177676e+07  133.876013   612.346769   
50%     810.000000   21999.000000  3.095774e+08  239.047878   873.452242   
75%     982.500000   38233.000000  6.553768e+08  344.245856  1175.047740   
max    1320.000000  100919.000000  2.069555e+09  560.313070  1396.991494   

       crop_area_ha     s_area_ha    m_area_ha    s_prod_tot    m_prod_tot  
count     59.000000     59.000000    59.000000  5.900000e+01  5.900000e+01  
mean    2502.770273   1985.838890   472.684662  5.779463e+05  4.848132e+05  
std     3060.248057   2866.138229   701.150863  9.541010e+05  7.658267e+05  
min    

Both datasets capture substantial variation in population, crop yields and output between districts and subcounties. If you look at the data by district however, it is obvious to see that yields of maize and sorghum vary widely across a range with some areas yielding significantly more than others. On subcounty level, data shows huge variations in the size of crop areas and production with some doing significantly better than others. These data provide insights into the diversity of agricultural performance in Karamoja, and are particularly useful for informing more targeted interventions.

## 5.2 Exploring Relationships

## 5.1 Relationship between population and maize yield

In [51]:
# Let us find the covariance between population and maize yield to see their relationship
cov_matrix = district_merged[['pop', 'm_yield_ha']].cov()
cov_population_maize_yield = cov_matrix.loc['pop', 'm_yield_ha']

print(f"Covariance between population and maize yield: {cov_population_maize_yield}")

Covariance between population and maize yield: -1994294.5952380951



There is a negative covariance of approximately -1,994,295 between population and maize yield which hereby suggests that there is an inverse relationship between them, as population increases, maize yield tends to decrease, or vice versa.

## 5.2 Relationship between population and sorghum yield

In [52]:
# Let us find the covariance matrix between the population and sorghum yield
cov_matrix_sorghum = district_merged[['pop', 's_yield_ha']].cov()

# Extract the covariance value between population and sorghum yield
cov_population_sorghum_yield = cov_matrix_sorghum.loc['pop', 's_yield_ha']

print(f"Covariance between population and sorghum yield: {cov_population_sorghum_yield}")

Covariance between population and sorghum yield: 548860.9761904756


There is a positive covariance of approximately 548,861 between population and sorghum yield this indicates that there is a direct relationship between them where as the population increases, sorghum yield tends to increase as well.

## 5.3 Relationship between Maize yield and Sorghum yield

In [53]:
# Let us find the covariance matrix between the population and sorghum yield
cov_matrix_maize_sorghum = district_merged[['m_yield_ha', 's_yield_ha']].cov()

# Extract the covariance value between maize yield and sorghum yield
cov_maize_sorghum_yield = cov_matrix_maize_sorghum.loc['m_yield_ha', 's_yield_ha']

print(f"Covariance between maize yield and sorghum yield: {cov_maize_sorghum_yield}")

Covariance between maize yield and sorghum yield: 21671.785714285714


There is a positive covariance of approximately 21,672 between maize yield and sorghum yield which indicates that higher maize yields are associated with higher sorghum yields. This shows that there is a direct relationship between them where an increase in maize yield is generally accompanied by an increases in sorghum yield.









## 5.4 Prepare Data for Tableau.

In [61]:
# Export the cleaned and merged datasets to CSV files that can easily be imported into Tableau.
district_merged.to_csv('district_merged.csv', index=False)
subcounty_merged.to_csv('subcounty_merged.csv', index=False)
merged_cyp.to_csv('merged_cyp.csv', index=False)

# 6.0 Conclusion and Recommendations

The objective of this project was to explore and evaluate crop yields in the Karamoja region, Uganda using an interactive visualization tool with integrated data sources. We tried to look at how productive the district and sub-counties were in terms of agricultural production. Key findings include:


**Covariance Analysis:** We found a positive covariance between maize yield and sorghum yield, so in areas with higher yields of one crop we also tends to find high values for the other. In contrast, the covariance coefficient between population and maize yield was negative because high er populations correspond with lower maize yields.


**Summary Statistics:** From the summary statistics, there is clearly variability in crop yields and productivity among districts (sub-counties) as indicated by different farmers who performed significantly differently with regards to how they are currently carrying out their agricultural activities.

RecommendationsThe results suggest the following strategies:

**Targeted Intervention in Agriculture:** Target agriculture support/intervention and input distribution to areas with low maize/sorghum yield. This will help in better and high yielding of overall crop productivity too.

**Design of Crop Planning Strategies based on Population Density:** Incorporate population density data in crop planning strategies. Priority needs to be given for sustainable crop management practices in areas having higher populations so as to achieve balance between agricultural output and the need of population.

**Advanced Spatial Analysis:** Make use of the maize and sorghum maps to extract spatial data on crop yields. It can enable efficient resource allocation and Local interventions to solve local problems.

These recommendations aim to enhance food security and improve agricultural productivity in Karamoja by leveraging the insights gained from our data analysis.